### PR TITLE
TINKERPOP-1437 Added tests for dedup(Scope,String...) overload in DedupTest

### PR DESCRIPTION
**Pull request by spmallette**
https://issues.apache.org/jira/browse/TINKERPOP-1437

Pretty simple - just added a couple of tests - but didn't CTR as I'd feel better with a couple of extra eyes on this as we've had bad tests slip into the suite in the past.  

Runs nicely with `mvn clean install` - doing a `docker/build.sh -t -i -n` now. If anything bad happens I'll post back here.

VOTE +1

### CODE REVIEW

File: gremlin-test/src/main/java/org/apache/tinkerpop/gremlin/process/traversal/step/filter/DedupTest.java [hunk 1]
```diff
@@ -241,6 +279,16 @@ public void g_V_both_both_dedup_byXoutE_countX_name() {
 
     public static class Traversals extends DedupTest {
         @Override
+        public Traversal<Vertex, String> get_g_V_out_in_valuesXnameX_fold_dedupXlocalX_unfold() {
+            return g.V().out().in().values("name").fold().dedup(Scope.local).unfold();
+        }
+
+        @Override
+        public Traversal<Vertex, Map<String, String>> get_g_V_out_in_asXxX_in_asXyX_selectXx_yX_byXnameX_fold_dedupXlocal_x_yX_unfold() {
```
Comment (okram) [inline gremlin-test/src/main/java/org/apache/tinkerpop/gremlin/process/traversal/step/filter/DedupTest.java]: You named the method wrong. Should be `get_g_V_out_asXxX_in_asXyX.....`

File: gremlin-test/src/main/java/org/apache/tinkerpop/gremlin/process/traversal/step/filter/DedupTest.java [hunk 2]
```diff
@@ -73,6 +77,40 @@
...
+        assertEquals(3, names.size());
+        assertThat(names, containsInAnyOrder("marko", "josh", "peter"));
+    }
+
+    @Test
+    @LoadGraphWith(MODERN)
+    public void g_V_out_in_asXxX_in_asXyX_selectXx_yX_byXnameX_fold_dedupXlocal_x_yX_unfold() {
+        final Traversal<Vertex, Map<String, String>> traversal = get_g_V_out_in_asXxX_in_asXyX_selectXx_yX_byXnameX_fold_dedupXlocal_x_yX_unfold();
+        final List<Map<String,String>> names = traversal.toList();
+        assertEquals(6, names.size());
```
Comment (okram) [inline gremlin-test/src/main/java/org/apache/tinkerpop/gremlin/process/traversal/step/filter/DedupTest.java]: You can use the `checkResult` and `MapMaker` to make this cleaner if you like. Check `WhereTest` for examples:

```
    @Test
    @LoadGraphWith(MODERN)
    public void g_V_asXaX_outEXcreatedX_asXbX_inV_asXcX_inXcreatedX_asXdX_whereXa_ltXbX_orXgtXcXX_andXneqXdXXX_byXageX_byXweightX_byXinXcreatedX_valuesXageX_minX_selectXa_c_dX() {
        final Traversal<Vertex, Map<String, String>> traversal = get_g_V_asXaX_outEXcreatedX_asXbX_inV_asXcX_inXcreatedX_asXdX_whereXa_ltXbX_orXgtXcXX_andXneqXdXXX_byXageX_byXweightX_byXinXcreatedX_valuesXageX_minX_selectXa_c_dX();
        printTraversalForm(traversal);
        checkResults(makeMapList(3,
                "a", "peter", "c", "lop", "d", "josh",
                "a", "peter", "c", "lop", "d", "marko",
                "a", "josh", "c", "lop", "d", "marko",
                "a", "josh", "c", "lop", "d", "peter"), traversal);
    }
```

File: gremlin-test/src/main/java/org/apache/tinkerpop/gremlin/process/traversal/step/filter/DedupTest.java [hunk 3]
```diff
@@ -73,6 +77,40 @@
...
+    public void g_V_out_in_valuesXnameX_fold_dedupXlocalX_unfold() {
+        final Traversal<Vertex, String> traversal = get_g_V_out_in_valuesXnameX_fold_dedupXlocalX_unfold();
+        final List<String> names = traversal.toList();
+        assertEquals(3, names.size());
+        assertThat(names, containsInAnyOrder("marko", "josh", "peter"));
+    }
+
+    @Test
+    @LoadGraphWith(MODERN)
+    public void g_V_out_in_asXxX_in_asXyX_selectXx_yX_byXnameX_fold_dedupXlocal_x_yX_unfold() {
```
Comment (okram) [inline gremlin-test/src/main/java/org/apache/tinkerpop/gremlin/process/traversal/step/filter/DedupTest.java]: This has a bad method name --- see my other comment and then fix it here too.

File: gremlin-test/src/main/java/org/apache/tinkerpop/gremlin/process/traversal/step/filter/DedupTest.java [hunk 4]
```diff
@@ -73,6 +77,40 @@
...
+
+    public abstract Traversal<Vertex, Map<String, String>> get_g_V_out_in_asXxX_in_asXyX_selectXx_yX_byXnameX_fold_dedupXlocal_x_yX_unfold();
+
+    @Test
+    @LoadGraphWith(MODERN)
+    public void g_V_out_in_valuesXnameX_fold_dedupXlocalX_unfold() {
+        final Traversal<Vertex, String> traversal = get_g_V_out_in_valuesXnameX_fold_dedupXlocalX_unfold();
+        final List<String> names = traversal.toList();
+        assertEquals(3, names.size());
+        assertThat(names, containsInAnyOrder("marko", "josh", "peter"));
```
Comment (okram) [inline gremlin-test/src/main/java/org/apache/tinkerpop/gremlin/process/traversal/step/filter/DedupTest.java]: Note we have `checkResults()`... e.g.

```
    @Test
    @LoadGraphWith(MODERN)
    public void g_V_whereXnotXoutXcreatedXXX_name() {
        final Traversal<Vertex, String> traversal = get_g_V_whereXnotXoutXcreatedXXX_name();
        printTraversalForm(traversal);
        checkResults(Arrays.asList("vadas", "lop", "ripple"), traversal);
    }
```

**Comment by okram** [comment]
Made a few comments to ensure your test code is consistent with the others. Besides minor cosmetics -- VOTE +1.

**Comment by dkuppitz** [comment]
VOTE: +1

Question: For future tests, should we always prefer Hamcrest over JUnit?

**Comment by spmallette** [comment]
I typically prefer JUnit for all simple assertions except for `assertTrue` and `assertFalse`. Those two really stink compared to hamcrest and `assertThat(something(), is(true))` or `assertThat(something(), is(false))` because they print better output than the former that makes it easier to immediately see the problem with the assertion on failure. I convert those `assertTrue` and `assertFalse` as I see opportunity to do so. I also use hamcrest when the assertion logic is complex and there is a better simple one liner operation with hamcrest that makes the assertion more readable with a cleaner output message than junit.  So, no, I don't think we have to prefer Hamcrest to Junit across the board, but we should consider it when it produces a more readable output or more readable code.

**Pull request merged**